ext/src/ruby_api/func.rs

```diff
@@ -198,36 +212,67 @@ impl From<BoxValue<Exception>> for InvokeError {
     }
 }
 
-struct Caller<'a> {
-    inner: &'a RefCell<CallerImpl<'a, StoreData>>,
+pub struct Caller<'a> {
+    inner: UnsafeCell<Option<CallerImpl<'a, StoreData>>>,
```

Comment:
Probably worth a comment as to why we use `UnsafeCell` here.

> We use `UnsafeCell` because we need interior mutability, but can't use dynamically checked `Refcell` because (... reasons? I assume GC?).

Reply:
After thinking about it more, it might be nice to wrap this functionality in a new type. That way, we can specify the behavior a bit more if we need to. At the very least, it can help to communicate the purpose a bit more clearly.

```rust
/// A handle to a [`Caller`] which allows for multiple re-entant borrows (i.e. Ruby -> Wasm -> Ruby -> Wasm).
pub struct CallerHandle<'a> {
  value: UnsafeCell<Option<CallerImpl<'a, StoreData>>>
}

impl <'a>CallerHandle<'a> {
  pub fn get_mut(&self) -> Somethin {
    // Add logic here to keep track of borrowers if possible...

    unsafe { &mut *self.inner.get() }
      .as_mut()
      .ok_or_else(|| error!("Caller outlived its Func execution"))?
  }

  pub fn expire(&self) {
    unsafe { *self.value.get() = None }
  }
}

Reply:
I'm using `UnsafeCell` because I couldn't get it to compile with `RefCell`, not because of re-entrance. Callers don't need to be re-entrant because you get a new caller for each entrance, that shouldn't be a problem.

The problem I ran into is that I couldn't find a way to return a `wasmtime::store::context::StoreContext<_>` from a `RefCell`:


```rust
pub struct CallerRefCell<'a> {
    inner: RefCell<Option<wasmtime::Caller<'a, StoreData>>>,
}

impl<'a> CallerRefCell<'a> {
    pub fn context(&self) -> Result<StoreContext<StoreData>, Error> {
        match *self.inner.borrow() {
            Some(ref inner) => Ok(inner.as_context()),
            None => err!("Caller outlived its Func execution"),
        }
    }

    pub fn context_mut(&self) -> Result<StoreContextMut<StoreData>, Error> {
        match *self.inner.borrow_mut() {
            Some(ref mut inner) => Ok(inner.as_context_mut()),
            None => err!("Caller outlived its Func execution"),
        }
    }
}
```


```
error[E0515]: cannot return value referencing temporary value
   --> ext/src/ruby_api/func.rs:315:32
    |
314 |         match *self.inner.borrow() {
    |                ------------------- temporary value created here
315 |             Some(ref inner) => Ok(inner.as_context()),
    |                                ^^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function

error[E0515]: cannot return value referencing temporary value
   --> ext/src/ruby_api/func.rs:322:36
    |
321 |         match *self.inner.borrow_mut() {
    |                ----------------------- temporary value created here
322 |             Some(ref mut inner) => Ok(inner.as_context_mut()),
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function
```

Not knowing how to fix this I resorted to `UnsafeCell`. I have a hunch that `context` and `context_mut` are actually unsafe: the context can become invalid if the caller is expired (set the Option to None), and the type system wouldn't know about that. In practice it is probably not a problem, though, as the Wasmtime's context will still exist.

Let me know if you can think of a way to avoid resorting to `UnsafeCell`.

Reply:
I used your recommendation of wrapping the UnsafeCell into a new type, but nothing keeps track of borrows. I also cherry-picked your RTypedData commit (bdd8486d4f5439ff149366fc6f00e21592255338).

I (again) looked into using `RefCell` and failed. Maybe it'd work by _not_ implementing `context` / `context_mut` on `StoreContextValue`, and instead getting the context from the enum variants directly when needed (with a macro?). But that'd be clunky, 😞.

Reply:
I'd like to see how you did it if you still have that commit around.

Reply:
I'm going to merge this to keep moving forward, but happy to iterate on it once in main. Any and all feedback is still welcome 🙇.